File: apps/runtime_daemon/descriptor.py
"""Private runtime descriptor schema and loading for the local daemon.

The descriptor is the CLI's only route into daemon-owned state: it carries the
loopback URL, process identity, and the local bearer token. Provider keys are
never stored here — they remain environment-resolved by the daemon process.
"""

from __future__ import annotations

import json
import os
import secrets
import uuid
from pathlib import Path
from typing import Literal

from pydantic import Field, ValidationError

from agent_os_contracts import (
    ContractModel,
    NonEmptyStr,
    SURFACE_PROTOCOL_VERSION,
    SurfaceProtocolVersion,
    SurfaceProtocolVersionError,
    UtcDateTime,
    canonical_json,
    parse_surface_protocol_version,
    surface_protocol_readable_versions,
)


class RuntimeDescriptorError(ValueError):
    """The runtime descriptor is missing, malformed, or unsafe."""


class RuntimeDescriptorProtocolError(RuntimeDescriptorError):
    """The descriptor names a surface protocol version this build does not read.

    A version skew, not a corrupt file: the descriptor is well-formed JSON with a
    well-formed ``MAJOR.MINOR`` version, and it is simply outside the range this
    build negotiates. Kept as its own type because the reaction differs — a skew
    means another build's daemon is probably RUNNING, so replacing its descriptor
    (which is its only handle) is the wrong move, while a genuinely unreadable
    file is not evidence of anything.
    """


class RuntimeDescriptor(ContractModel):
    protocol_version: SurfaceProtocolVersion
    pid: int = Field(gt=0)
    boot_id: NonEmptyStr
    host: Literal["127.0.0.1"]
    port: int = Field(ge=1, le=65535)
    bearer_token: NonEmptyStr
    database_path: NonEmptyStr
    workspace_path: NonEmptyStr
    created_at: UtcDateTime

    @property
    def base_url(self) -> str:
        return f"http://{self.host}:{self.port}"


def generate_runtime_token() -> str:
    """One random 256-bit local bearer token (url-safe base64)."""

    return secrets.token_urlsafe(32)


def generate_boot_id() -> str:
    """One unique daemon boot identity."""

    return f"boot:{uuid.uuid4()}"


def load_runtime_descriptor(path: Path) -> RuntimeDescriptor:
    """Parse and validate one private runtime descriptor file.

    A descriptor written by another build is reported as the version skew it is
    rather than as a schema failure: the same ``MAJOR.MINOR`` that this build
    refuses to speak is what tells an operator which side to upgrade.
    """

    try:
        raw = path.read_text(encoding="utf-8")
    except OSError as exc:
        raise RuntimeDescriptorError(
            f"cannot read runtime descriptor {path}"
        ) from exc
    try:
        value = json.loads(raw)
    except json.JSONDecodeError as exc:
        raise RuntimeDescriptorError(
            f"runtime descriptor {path} is not valid JSON"
        ) from exc
    if not isinstance(value, dict):
        raise RuntimeDescriptorError("runtime descriptor must be a JSON object")
    declared = value.get("protocol_version")
    if isinstance(declared, str):
        try:
            parse_surface_protocol_version(declared)
        except SurfaceProtocolVersionError:
            # Not a version at all; the schema error below is the honest report.
            pass
        else:
            readable = surface_protocol_readable_versions(SURFACE_PROTOCOL_VERSION)
            if declared not in readable:
                raise RuntimeDescriptorProtocolError(
                    f"runtime descriptor {path} speaks surface protocol {declared}, "
                    f"which this build does not read (reads {', '.join(readable)}); "
                    "this is a version skew, not a corrupt descriptor — upgrade the "
                    "older side, or delete the descriptor if no daemon is running"
                )
    try:
        return RuntimeDescriptor.model_validate(value)
    except ValidationError as exc:
        raise RuntimeDescriptorError(
            f"runtime descriptor {path} has an invalid schema"
        ) from exc


def write_descriptor(path: Path, descriptor: RuntimeDescriptor) -> None:
    """Atomically persist one private descriptor at mode 0600."""

    path.parent.mkdir(mode=0o700, parents=True, exist_ok=True)
    if path.exists() and path.is_symlink():
        raise RuntimeDescriptorError("runtime descriptor cannot be a symlink")
    temporary = path.with_name(path.name + f".{descriptor.boot_id}.tmp")
    flags = os.O_WRONLY | os.O_CREAT | os.O_EXCL
    fd = os.open(temporary, flags, 0o600)
    try:
        os.write(fd, canonical_json(descriptor).encode("utf-8"))
        os.fsync(fd)
    finally:
        os.close(fd)
    os.replace(temporary, path)

File: apps/cli-ts/src/descriptor.ts
/**
 * Private runtime descriptor loading (~/.agent-os/runtime.json by default).
 * The descriptor is the client's only route into daemon-owned state; the
 * bearer token never leaves this process and is never logged.
 */
import { readFile } from "node:fs/promises";
import { homedir, hostname } from "node:os";
import { join } from "node:path";
import { z } from "zod";

import {
  SURFACE_PROTOCOL_READABLE_VERSIONS,
  SURFACE_PROTOCOL_VERSION,
  SurfaceProtocolVersionSchema,
} from "./contracts.js";

export const DEFAULT_RUNTIME_DESCRIPTOR = join(homedir(), ".agent-os", "runtime.json");

/** Shape of a `MAJOR.MINOR` version, mirroring `parse_surface_protocol_version`. */
const PROTOCOL_VERSION_SHAPE = /^(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)$/;

/** The path holds no descriptor: there is no daemon to attach to. */
export class RuntimeDescriptorNotFoundError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "RuntimeDescriptorNotFoundError";
  }
}

/**
 * The descriptor is readable but names a surface protocol version this client
 * does not read.
 *
 * Typed because the reaction is not "start a daemon": the descriptor was written
 * by another build, which is probably still RUNNING, and this file is that
 * daemon's only handle. Replacing it would strand a live daemon; the caller must
 * stop and say so instead.
 */
export class SurfaceProtocolSkewError extends Error {
  constructor(
    message: string,
    readonly clientVersion: string,
    readonly serverVersion: string,
  ) {
    super(message);
    this.name = "SurfaceProtocolSkewError";
  }
}

const RuntimeDescriptorSchema = z.object({
  // Ordered acceptance, not a pin: a descriptor written by a runtime one minor
  // behind still names a version this client reads (see SURFACE_PROTOCOL_READABLE_VERSIONS).
  protocol_version: SurfaceProtocolVersionSchema,
  pid: z.number().int().positive(),
  boot_id: z.string().min(1),
  host: z.literal("127.0.0.1"),
  port: z.number().int().min(1).max(65535),
  bearer_token: z.string().min(1),
  database_path: z.string().min(1),
  workspace_path: z.string().min(1),
  created_at: z.string().min(1),
});

export type RuntimeDescriptor = z.infer<typeof RuntimeDescriptorSchema> & {
  baseUrl: string;
};

export async function loadRuntimeDescriptor(
  path: string = DEFAULT_RUNTIME_DESCRIPTOR,
): Promise<RuntimeDescriptor> {
  let raw: string;
  try {
    raw = await readFile(path, "utf8");
  } catch {
    throw new RuntimeDescriptorNotFoundError(
      `runtime descriptor not found at ${path}; start the daemon first (noem daemon start / scripts/dev_daemon.py)`,
    );
  }
  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch {
    throw new Error(`runtime descriptor at ${path} is not valid JSON`);
  }
  const declared =
    typeof parsed === "object" && parsed !== null
      ? (parsed as { protocol_version?: unknown }).protocol_version
      : undefined;
  if (
    typeof declared === "string" &&
    PROTOCOL_VERSION_SHAPE.test(declared) &&
    !(SURFACE_PROTOCOL_READABLE_VERSIONS as readonly string[]).includes(declared)
  ) {
    // Reported as the version skew it is, before the schema dump: a well-formed
    // version outside the declared set is a build mismatch, and the operator
    // needs to know which side is older, not which field zod disliked.
    throw new SurfaceProtocolSkewError(
      `runtime descriptor at ${path} speaks surface protocol ${declared}, which this ` +
        `client ${SURFACE_PROTOCOL_VERSION} does not read (reads ` +
        `${SURFACE_PROTOCOL_READABLE_VERSIONS.join(", ")}); this is a version skew, not ` +
        `a corrupt descriptor — upgrade the older side, or delete the descriptor if no ` +
        `daemon is running`,
      SURFACE_PROTOCOL_VERSION,
      declared,
    );
  }
  const descriptor = RuntimeDescriptorSchema.parse(parsed);
  return { ...descriptor, baseUrl: `http://${descriptor.host}:${descriptor.port}` };
}

export function localHostname(): string {
  return hostname() || "local";
}
